File: src/main/java/org/academy/api/client/gui/widget/FillWidget.java
package org.academy.api.client.gui.widget;

import org.academy.api.client.gui.drawable.ColorDrawable;
import org.academy.api.client.gui.drawable.Drawable;
import org.academy.api.client.gui.render.RenderContext;
import org.jspecify.annotations.Nullable;

public class FillWidget extends AbstractWidget {
    public FillWidget(int color) {
        setBackground(new ColorDrawable(color));
    }

    @Override
    public void render(RenderContext context) {
        if (!isVisible()) return;
        super.render(context);
    }

    public int getColor() {
        if (getBackground() instanceof ColorDrawable colorDrawable) {
            return colorDrawable.getColor();
        }
        return 0;
    }

    @Override
    public void setBackground(@Nullable Drawable background) {
        if (background instanceof ColorDrawable) super.setBackground(background);
    }

    public FillWidget setColor(int color) {
        if (getBackground() instanceof ColorDrawable colorDrawable) {
            colorDrawable.setColor(color);
        } else {
            setBackground(new ColorDrawable(color));
        }
        return this;
    }
}

File: src/main/kotlin/org/academy/api/client/gui/widget/FillWidget.kt
package org.academy.api.client.gui.widget

import org.academy.api.client.gui.drawable.ColorDrawable
import org.academy.api.client.gui.drawable.Drawable
import org.academy.api.client.gui.render.RenderContext

class FillWidget(color: Int) : AbstractWidget() {
    override var background: Drawable? = null
        set(background) {
            if (background is ColorDrawable) field = background
        }

    init {
        background = ColorDrawable(color)
    }

    override fun render(context: RenderContext) {
        if (!isVisible()) return
        super.render(context)
    }

    val color: Int
        get() {
            val bg = background
            if (bg is ColorDrawable) {
                return bg.color
            }
            return 0
        }

    fun setColor(color: Int): FillWidget {
        val bg = background
        if (bg is ColorDrawable) {
            bg.color = color
        } else {
            background = ColorDrawable(color)
        }
        return this
    }
}